### DISCUSSION TITLE
v4.0.34 pre-release

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#63] [General]
I've released v4.0.34 as a pre-release.

Brilliant feedback received for v4.0.33 (https://github.com/BJReplay/ha-solcast-solar/discussions/60), and this incorporates that feedback, plus resolves other known issues.

Many thanks to @isorin for his brilliant attention to a dampening-related issue. I might have tipped in a little bit as well for other things in this release.

Thanks also to @gcoan, @Noodleyman, @ProphetOfDoom and @Nilogax for their testing reports.

To give this pre-release a go, please make sure that ```Show beta versions``` is on when re-downloading from HACS.

As always, both positive and negative feedback is most welcome, and this will contribute to getting the release ultimately published as latest. Report back, and don't be shy!

The BJReplay fork is still not published for searching in HACS, so continued use of a manual install in HACS is the go. As explained earlier we're going to keep any releases as pre-release on the fork until that happens, as any 'latest' version change would require modification of that HACS pull request.

## What's Changed

* Fix query_forecast_data so that near-term historical forecasts are returned by @isorin
* Instantly fall-back to cache on reload if rooftop/usage API calls fail, which can reduce start time by @autoSteve (fixes a known issue)
* An async call timeout of sites get will fall back to cache if it exists by @autoSteve
* Much logging improvements by @autoSteve
* Sites cache being sometimes incorrectly created with the API key appended, despite only having one API key by @autoSteve
* Redaction of latitude/longitude in debug logs by @autoSteve
* Likely elimination of 'tally' warnings by @autoSteve (should fix a known issue)
* Fix API usage retry mechanism by @autoSteve (fixes a known issue)

Full Changelog: https://github.com/BJReplay/ha-solcast-solar/compare/v4.0.33...v4.0.34

## Known issues

None

**Comment by ProphetOfDoom** [comment]
Before I update to V4.0.34, I thought I'd just mention/confirm that I currently have both a solcast-usage.json file and a solcast-usage-apikey.json file. The former was last updated at 01:00 BST and the latter when my automation ran at approx 04:49 BST.
Diving in now ...

PS. As I only have the one api key, I'll delete the solcast-usage-apikey.json before proceeding

PPS. Now upgraded to V4.0.34. No errors in log. All sensors and roof tallys populated. solcast-sites.json and solcast-usage.json files updated. No extraneous solcast-sites-apikey.json file created (as expected)

PPPS. V4.0.34. solcast_solar.update_forecasts service called. API calls updated from 2/10 to 4/10. solcast.json and solcast-usage.json files updated. All sensores populated and roof tally values match values in solcast.json file. No solcast related errors in log

**Comment by Nilogax** [comment]
Updated this afternoon, no errors in the log and scheduled forecast updates working OK. History retained back as far my 0.33 update. So far so good!

**Reply by isorin** [reply]
@Nilogax Btw. if you delete the history, restart and trigger a forecast update, it should bring back forecasts for the past 7 days at the expense of 1 API use (plus 8 days in the future as usual)
I forgot about this ...
Actually just a restart after deleting '/config/solcast.json' wil trigger the API call. No need to do a forecast update.

**Reply by autoSteve** [reply]
Technically two API uses. One for the seven day history, then another for eight days future.

Taking one for the team in terms of API use, I just re-tested it. Worked as advertised, @Nilogax.

Now restoring from backup to restore the rest of history...

**Reply by Nilogax** [reply]
Thanks @isorin - I'll give that a go later.

**Reply by Nilogax** [reply]
Done! Recent history is now populated so I'll turn my dampening gubbins back on to see if anything breaks. Thanks again @isorin!

**Comment by gcoan** [comment]
Updated just now to 4.0.34 beta.

Installed OK, HA restarted OK and no visible errors in the logfile.

Sensors checked and they all look OK, no drops or missing data.

Out of API calls at the moment so can't do a poll until after midnight.

One minor thing, can this be suppressed/removed:

```
2024-06-28 22:55:38.430 DEBUG (MainThread) [custom_components.solcast_solar] Solcast Migrating from version 7
2024-06-28 22:55:38.432 DEBUG (MainThread) [custom_components.solcast_solar] Solcast Migration to version 7 successful
```

Whatever version 7 was was some time ago, the migration (if needed) has already been done so why am I getting messages about it still?

Always nice to find things for @autoSteve to add to his to-do list.  I added a new FR #64 as well today 😉

**Reply by gcoan** [reply]
Further update:

Just looking at the sensors and I can see the issue with day 7 forecast dropping to zero at midnight and then resuming when the first poll had occurred has been resolved, the sliding time window is working perfectly and no drops to zero.

Scheduled poll worked perfectly, no retries, everything in the log looked good

**Reply by autoSteve** [reply]
Would this please you, Geoffrey?

```
2024-06-29 11:15:32.982 INFO (MainThread) [homeassistant.setup] Setting up solcast_solar
2024-06-29 11:15:32.983 INFO (MainThread) [homeassistant.setup] Setup of domain solcast_solar took 0.00 seconds
2024-06-29 11:15:32.983 DEBUG (MainThread) [custom_components.solcast_solar] SOLCAST - Options version 7
2024-06-29 11:15:32.990 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Sites cache exists
2024-06-29 11:15:32.990 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - connecting to - https://api.solcast.com.au/rooftop_sites?format=json&api_key=******NV8ggJ
```

What is happening in ```__init__``` is that as time goes by the integration configuration options can get changed, and the code uses a versioning system to keep track of this. The existing code was a bit sloppy, but functional.

I will make a change in 4.0.next that will state the options version on startup at debug level, then will further log at the same level should a change to available options actually be needed.

**Reply by gcoan** [reply]
thanks, it was the bit that said it was migrating from version 7 every time I found misleading.  If there's inbuilt versioning of config and it wants to track that and debug log it, fine

**Reply by autoSteve** [reply]
PR #65.

**Reply by autoSteve** [reply]
I think Github just had a stroke... 😂

<img width="380" alt="image" src="https://github.com/BJReplay/ha-solcast-solar/assets/37229860/aab77866-7a27-4df0-87e7-5b929db21b29">
